File: online-training-platform/leetcode/longest-substring-without-repeating-characters.kt
 class Solution {
    fun lengthOfLongestSubstring(s: String): Int {
      var start = 0
      var end = 0
      var res = 0
      val map = HashMap<Char, Int>()
      
      while( end < s.length){
        if(map.contains(s[end])){
          val newStart = map[s[end]]!! + 1
          start = Math.max(start, newStart)
        }
        map[s[end]] = end
        res = Math.max(res, end - start + 1)
        end += 1
      }
      
      return res
    }
}

File: online-training-platform/leetcode/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int left = 0, right = 0, res = 0, n = s.size();
        vector<int> letters(128,0);
        while(right < n) {
            letters[s[right]]++;
            
            while(letters[s[right]] > 1) {
                letters[s[left]]--;
                left++;
            }
            
            res = max(res, right - left + 1);
            right++;
        }
        
        return res;
    }
};